feat: Rule Performance Statistics for flat ESLint

mnkiefer opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

- Introduces a new flag `--stats` to flat ESLint, which enables the user to obtain a series of runtime statistics on top of their final lint results (refer to the table below).
- See [RFC](https://github.com/eslint/rfcs/blob/main/designs/2023-rule-performance-statistics/README.md)

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

nzakas commented:
Just converting this to a draft so we know it's not yet time to review it.

bmish commented on lib/shared/stats.js:
```diff
@@ -0,0 +1,81 @@
+    /**
+     * End time measurement
+     * @param {float} t Variable for tracking time
+     * @returns {void}
```
void?

mnkiefer replied:
This should be something like this?
```suggestion
     * @returns {float} The measured time in milliseconds
```

bmish replied:
Yep.

bmish commented on lib/linter/linter.js:
```diff
@@ -91,6 +92,7 @@ const parserSymbol = Symbol.for("eslint.RuleTester.parser");
  * @property {SourceCode|null} lastSourceCode The `SourceCode` instance that the last `verify()` call used.
  * @property {SuppressedLintMessage[]} lastSuppressedMessages The `SuppressedLintMessage[]` instance that the last `verify()` call produced.
  * @property {Map<string, Parser>} parserMap The loaded parsers.
+ * @property {Object} times The times spent on applying a rule to a file (see `stats` option).
```
Can we define more specific types instead of using `Object` and `Map` in some of the jsdocs?

mnkiefer replied:
Of course! 👍  Please see the suggestion below as well as the others from my review.
```suggestion
 * @property {Times} times The times spent on applying a rule to a file (see `stats` option).
```

mnkiefer commented on lib/shared/types.js:
```diff
@@ -189,11 +189,32 @@ module.exports = {};
+ * @typedef {Object} Times
+ * @property {number} fix The time for the fix pass.
+ * @property {number} parse The time that is spent when parsing a file.
+ * @property {Object} rules Object of times for each rule.
```
```suggestion
 * @property {Object<string, number>} rules Object of times for each rule.
```

bmish replied:
That may be fine although I believe `Record<...>` is more common.

mnkiefer commented on lib/linter/linter.js:
```diff
@@ -963,9 +966,11 @@ const BASE_TRAVERSAL_CONTEXT = Object.freeze(
  * @param {string | undefined} cwd cwd of the cli
  * @param {string} physicalFilename The full path of the file on disk without any code block information
+ * @param {boolean} stats If true, stats are collected appended to the result
+ * @param {Map} slots InternalSlotsMap of linter
```
```suggestion
 * @param {WeakMap<Linter, LinterInternalSlots>} slots InternalSlotsMap of linter
```

mnkiefer commented on lib/linter/timing.js:
```diff
@@ -128,20 +130,28 @@ module.exports = (function() {
      * @param {any} key key from the data object
      * @param {Function} fn function to be called
+     * @param {string} filename current (physical) filename
+     * @param {Map} slots linter internal slots map
```
```suggestion
     * @param {WeakMap<Linter, LinterInternalSlots>} slots linter internal slots map
```

mnkiefer commented on lib/shared/stats.js:
```diff
@@ -0,0 +1,81 @@
+     * Store time measurements in map
+     * @param {float} time Time measurement
+     * @param {Object} timeOpts Options relating which time was measured
+     * @param {Map} slots Linter internal slots map
```
```suggestion
     * @param {WeakMap<Linter, LinterInternalSlots>} slots Linter internal slots map
```

nzakas commented:
@me4502 getting back around to this now. Can you please rebase on top of main? And if you feel you're ready for review now, please feel free to hit that button. :)

mnkiefer commented:
> Getting back around to this now. Can you please rebase on top of main? And if you feel you're ready for review now, please feel free to hit that button. :)

Great to hear! I just did a rebase, so the implementation should be up to date with the new ESLint now. All that is left is the tests, but I suppose those would come in a second review round? So, pressing the Review button now 👍

nzakas commented:
Go ahead and add the tests now. We like to review the code and the tests together as oftentimes the tests tell more of the story than the code does.

nzakas commented on lib/linter/linter.js:
```diff
@@ -2016,7 +2016,7 @@ class Linter {
 
     getFixPasses() {
-        return internalSlotsMap.get(this).fixPasses;
+        return internalSlotsMap.get(this).fixPasses || 0;
```
Would it be more appropriate to use `??` here? (I'm assuming `fixPasses` is either missing or a number.)

nzakas commented on tests/lib/eslint/eslint.js:
```diff
@@ -4885,6 +4885,71 @@ describe("ESLint", () => {
+        }
+
+        it("should report stats", async () => {
+            process.chdir(originalDir);
```
Because you are doing this in both test cases, it would be better to move this into a `beforeEach`.

nzakas commented on lib/linter/linter.js:
```diff
@@ -935,9 +939,12 @@ function createRuleListeners(rule, ruleContext) {
  */
-function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageOptions, settings, filename, disableFixes, cwd, physicalFilename, ruleFilter) {
+function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageOptions, settings, filename, disableFixes, cwd, physicalFilename, ruleFilter,
+    stats, slots) {
```
It looks like we end up passing `slots` around unnecessarily (from here into timing.js and into stats.js). Can we just pull out the necessary values and pass them along?

(`slots` is really not supposed to be shared outside of `Linter`.)

mnkiefer replied:
Yes, this seems a bit redundant now that you mention it. ☺️ I've just updated the code accordingly:
- Removed `slots` from _timing.js_ `time()`
- Moved `storeTime()` to the Linter, since:
  - It is not used anywhere else
  - It's job is to store/update the time delta/ fixPasses in the Linter's slots map
  - Hence, `slots` is ok to use as input here (as we are inside the Linter).

nzakas commented on lib/shared/stats.js:
```diff
@@ -0,0 +1,81 @@
+"use strict";
```
Please a file header comment here (see other files for an example).

nzakas commented on lib/shared/stats.js:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * Start time measurement
+     * @returns {float} t variable for tracking time
```
`float` isn't a value that things like VS Code will recognize. Can we change these to `number`?

mnkiefer replied:
Done! See: https://github.com/eslint/eslint/pull/17850/commits/2d958121bde032fbfe95d24f0058237f7ba01538, https://github.com/eslint/eslint/pull/17850/commits/b671de4a75009d2ff19f9686fd40c08b60e64b45

nzakas reviewed (changes requested):
This is looking good. I'd just like to avoid passing `stats` around in so many places. (I left a comment inline about this, as well.) If we can instead pull that parts out of `slots` and later assign them back, we can make the additional functionality more testable.

nzakas commented on lib/linter/linter.js:
```diff
@@ -1970,6 +2069,14 @@ class Linter {
     }
 
+    getTimes(filename) {
+        return internalSlotsMap.get(this).times[filename] || { passes: [] };
```
Maybe another spot where `??` is more appropriate?
```suggestion
        return internalSlotsMap.get(this).times[filename] ?? { passes: [] };
```

nzakas commented on lib/linter/linter.js:
```diff
@@ -2043,9 +2150,12 @@ class Linter {
         const debugTextDescription = options && options.filename || `${text.slice(0, 10)}...`;
         const shouldFix = options && typeof options.fix !== "undefined" ? options.fix : true;
+        const stats = options && typeof options.stats !== "undefined" ? options.stats : false;
+        const filename = options && typeof options.filename !== "undefined" ? options.filename : "TEXT";
```
We tend to use `"<text>"` for this elsewhere in the codebase. Will that work here?

Also, a bit of cleanup similar to the previous line.

```suggestion
        const filename = options?.filename ?? "<text>";
```

nzakas commented on lib/linter/linter.js:
```diff
@@ -2043,9 +2150,12 @@ class Linter {
             currentText = text;
         const debugTextDescription = options && options.filename || `${text.slice(0, 10)}...`;
         const shouldFix = options && typeof options.fix !== "undefined" ? options.fix : true;
+        const stats = options && typeof options.stats !== "undefined" ? options.stats : false;
```
I think we can simplify a bit. Based on usage, I think it's fine to leave `stats` as `undefined` rather than setting to `false`, but let me know if I'm missing something.
```suggestion
        const stats = options?.stats;
```

nzakas commented on lib/shared/stats.js:
```diff
@@ -0,0 +1,41 @@
+     * @param {string} filename File that is currently being linted
+     * @returns {Object} Statistics result object
+     */
+    getStats(linter, filename) {
```
It might be worth passing in just the fix passes and times as separate arguments rather than passing in `linter`. This would make it easier to test this method.

mnkiefer replied:
You are right! 👍 Looking at it now after the refactoring, I also realized that we no longer need the `getStats()` function and just call:
```js
result.stats = {
    times: linter.getTimes(filePath),
    fixPasses: linter.getFixPasses()
};
```
See https://github.com/eslint/eslint/pull/17850/commits/762c74dcd2e97070ac76dbe5120115fb33094c8f, where I've also updated the header comment to account for this.

nzakas reviewed (changes requested):
The refactoring around `slots` looks good to me. Left a couple of suggestions for areas to clean up, but this is looking really good.

mdjermanovic commented on tests/lib/eslint/eslint.js:
```diff
@@ -4885,6 +4885,73 @@ describe("ESLint", () => {
+            assert.strictEqual(isNumber(results[0].stats.times.passes[1].total), true);
+        });
+
+    });
```
Can you also add some tests with [processors](https://eslint.org/docs/latest/extend/custom-processors)? I just tried `--stats` with [`eslint-plugin-markdown`](https://github.com/eslint/eslint-plugin-markdown) and the result doesn't seem to contain any parsing and rules timings data. Perhaps the filename-based logic doesn't account for this case where we're linting virtual child files.

mnkiefer replied:
@mdjermanovic: I've just added a test for processors. 👍

Doing so, I noticed that the lint times were only contributed to the results when using `lintFiles()` but not with `lintText()`. This was the result of not using the fully resolved filePath (for [placeholders](https://github.com/eslint/eslint/pull/17850/commits/439a7a75234a8f7b548567a90531c41150e8140a#diff-cb681134e1d583c87b8a91a2dfbd4de0bfe5dd84fa3ee57efb5028d7bfeb5219R494)) in my `getTimes()` helper function. to retrieve the respective `times` object.

Could you please try again and let me know whether you are still missing any timings data?

nzakas reviewed (dismissed):
Latest changes LGTM. Waiting for @mdjermanovic to re-review.

mnkiefer commented:
> Latest changes LGTM. Waiting for @mdjermanovic to re-review.

@nzakas: Any updates on this? 😇

mdjermanovic commented:
@mnkiefer sorry for the delay, I'll review it soon.

mdjermanovic commented:
Totals for rules should be the same as with [`TIMING`](https://eslint.org/docs/latest/extend/custom-rules#profile-rule-performance)? I'm getting lower numbers with `--stats` than with `TIMING`.

For example:

```js
// eslint.config.js
module.exports = [{
    languageOptions: {
        sourceType: "commonjs"
    },
    linterOptions: {
        reportUnusedDisableDirectives: "off"
    },
    rules: {
        indent: ["error", 4, { SwitchCase: 1 }]
    }
}];
```

```
$ TIMING=all node bin/eslint tests/lib/linter/linter.js --stats -f json
(node:7892) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)
[{"filePath":"C:\\projects\\tmp\\mnkiefer-eslint\\tests\\lib\\linter\\linter.js","messages":[],"suppressedMessages":[],"errorCount":0,"fatalErrorCount":0,"warningCount":0,"fixableErrorCount":0,"fixablea 
rningCount":0,"stats":{"times":{"passes":[{"parse":{"total":251.2441},"rules":{"indent":{"total":182.5374}},"fix":{"total":0},"total":619.8087}]},"fixPasses":0},"usedDeprecatedRules":[{"ruleId":"indent","replacedBy":[]}]}]
Rule   | Time (ms) | Relative
:------|----------:|--------:
indent |   214.375 |   100.0%
```

Per the stats, `indent` rule takes 182.5374ms, but TIMING in the same run outputs 214.375ms.

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -1057,7 +1116,15 @@ function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageO
+                    const ruleListenerResult = stats ? ruleListenerReturn.result : ruleListenerReturn;
+
+                    if (stats) {
+                        storeTime(ruleListenerReturn.tdiff, { filename: sharedTraversalContext.getPhysicalFilename(), type: "rules", key: ruleId }, slots, slots.resetTimes === filename);
```
Perhaps `slots.resetTimes === filename` is why the numbers are lower than expected, looks like each listener call resets time?

mdjermanovic replied:
Since the stats are now stored in the Linter, maybe Linter doesn't need to track them per filename, but only for the last run? That way, stats could not accumulate for multiple runs on the same file.

There are already methods that return data from the last run only, like `Linter#getSuppressedMessages()` and the newly added `Linter#getFixPasses()`.

So, `Linter#getStats()` would be returning stats from the last run. The stats would be reset on a new run, that is when `Linter#verify()` or `Linter#verifyAndFix()` is called.

mnkiefer replied:
@mdjermanovic: I just changed the `resetTimes` conditions. Now, if a file is parsed *again* the rules timing for that file will be reset. 👍 The *rules* output is now as expected and matches that of TIMING:
```
TIMING=all node bin/eslint tests/lib/linter/linter.js --stats -f json
[{"filePath":"/Users/I534083/Public/eslint-fork/tests/lib/linter/linter.js",
...
"stats":{"times":{"passes":[{
  "parse":{"total":179.162125},
  "rules":{"indent":{"total":140.72399700002433}},
  "fix":{"total":0},"total":402.135417}]},
  "fixPasses":0},
...
}]
Rule   | Time (ms) | Relative
:------|----------:|--------:
indent |   140.724 |   100.0%
```

mdjermanovic replied:
I still see the same version, maybe you didn't push the changes?

mnkiefer replied:
I just pushed them 👍 Please have a look now.

nzakas replied:
> I think we should simply reset stats at the start of verify/verifyAndFix, and I also think there's no need to keep stats per filename in Linter.

:+1:

mnkiefer replied:
@mdjermanovic: I agree, this also simplifies things nicely! 👍 I've just updated the code - could you please have another look?

mnkiefer commented:
@nzakas: Regarding the CLA, I tried to drop the two commits ('Update README') coming from a bot and somehow it pulled in all of the renovate changes. Could you please help to **revert this** as this seemed to have also pushed everyone onto the CLA.

nzakas commented:
@mnkiefer we can't actually revert those changes for you. We know you've already signed the CLA, so don't worry about the check.

In the future, though, it's a better idea to rebase on top of main rather than trying to merge with main:

```
git fetch
git rebase origin/main
```

mnkiefer commented:
> We know you've already signed the CLA, so don't worry about the check.

@nzakas, @mdjermanovic: FYI, I cleaned and put all of my changes into [a single commit](https://github.com/eslint/eslint/pull/17850/commits/e79b53570b73b9fd88705900f90216babf7a3e4f). The CLA check then get retriggered and everything is fine again. 👍

mnkiefer commented on lib/shared/types.js:
```diff
@@ -189,11 +189,32 @@ module.exports = {};
+ * @property {Times} times The times spent on (parsing, fixing, linting) a file.
+ */
+
+/**
+ * Performance Times for each ESLint pass
+ * @typedef {Times[]} Time passes
+ */
+
+/**
+ * @typedef {Object} Times
+ * @property {number} fix The time for the fix pass.
+ * @property {number} parse The time that is spent when parsing a file.
+ * @property {Record<string, number>} rules Object of times for each rule.
+ * @property {number} total The total time that is spent on (parsing, fixing, linting) a file.
+ */
```
@mdjermanovic: Yes, this was still outdated. I've just updated it, explicitly typing parse, rule, and fix times objects so they can easily be extended later (currently, they only contain the `total` time). What do you think? 

```suggestion
/**
 * Performance Times for each ESLint pass
 * @typedef {TimePass[]} Time passes
 */

/**
 * @typedef {Object} TimePass
 * @property {ParseTime} parse The parse object containing all parse time information.
 * @property {Record<string, RuleTime>} rules The rules object containing all lint time information for each rule.
 * @property {FixTime} fix The parse object containing all fix time information.
 * @property {string} total The total time that is spent on (parsing, fixing, linting) a file.
 */

/**
 * @typedef {Object} ParseTime
 * @property {number} total The total time that is spent when parsing a file.
 */

/**
 * @typedef {Object} RuleTime
 * @property {number} total The total time that is spent on a rule.
 */

/**
 * @typedef {Object} FixTime
 * @property {number} total The total time that is spent on a fix pass.
 */
```

mnkiefer replied:
@mdjermanovic: Please let me know if the updated [type definitions](https://github.com/eslint/eslint/pull/17850/files#diff-28a93c0c2a90d26ab4f8007aed5c72473b5650f25dde18060bac60c1081a2767R206-R227) are fine now.

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -1090,7 +1129,14 @@ function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageO
+        const ruleListeners = stats ? ruleListenersReturn.result : ruleListenersReturn;
+
+        if (stats) {
+            storeTime(ruleListenersReturn.tdiff, { filename, type: "rules", key: ruleId }, slots);
```
```suggestion
            storeTime(ruleListenersReturn.tdiff, { type: "rules", key: ruleId }, slots);
```

There's no need to pass `filename` anymore?

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -1100,7 +1146,15 @@ function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageO
+                    const ruleListenerResult = stats ? ruleListenerReturn.result : ruleListenerReturn;
+
+                    if (stats) {
+                        storeTime(ruleListenerReturn.tdiff, { filename, type: "rules", key: ruleId }, slots);
```
```suggestion
                        storeTime(ruleListenerReturn.tdiff, { type: "rules", key: ruleId }, slots);
```

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2152,15 +2247,46 @@ class Linter {
+                    fixPasses++;
+                    const time = endTime(t);
+
+                    storeTime(time, { filename, type: "fix" }, slots);
```
```suggestion
                    storeTime(time, { type: "fix" }, slots);
```

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2152,15 +2247,46 @@ class Linter {
+                    storeTime(time, { filename, type: "fix" }, slots);
+                    slots.fixPasses = fixPasses;
+                } else {
+                    storeTime(0, { filename, type: "fix" }, slots);
```
```suggestion
                    storeTime(0, { type: "fix" }, slots);
```

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -1335,12 +1387,26 @@ class Linter {
 
+            if (options.stats) {
+                const time = endTime(t);
+                const timeOpts = { type: "parse", filename: options.filename };
+
+                storeTime(time, timeOpts, slots);
```
```suggestion

                storeTime(time, { type: "parse" }, slots);
```

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -1619,12 +1687,26 @@ class Linter {
 
+            if (options.stats) {
+                const time = endTime(t);
+                const timeOpts = { type: "parse", filename: options.filename };
+
+                storeTime(time, timeOpts, slots);
```
```suggestion

                storeTime(time, { type: "parse" }, slots);
```

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2152,15 +2247,46 @@ class Linter {
             fixedResult = SourceCodeFixer.applyFixes(currentText, messages, shouldFix);
 
+            if (stats) {
+                t = startTime();
+                if (fixedResult.fixed) {
+                    fixPasses++;
+                    const time = endTime(t);
```
Looks like `time` doesn't track time spent on fixing the code, perhaps `t = startTime();` should be on line 2272?

mnkiefer replied:
You are right, the idea is the same (as for *parse*, etc.). We need to start the timer before `SourceCodeFixer.applyFixes()` for the *fix* time to be tracked.

mnkiefer commented on lib/linter/linter.js:
```diff
@@ -818,6 +822,38 @@ function stripUnicodeBOM(text) {
+    }
+
+    if (key) {
+        slots.times.passes[passIndex][type][key] = slots.times.passes[passIndex][type][key] || { total: 0 };
```
Thanks for this, I appreciate it! ☺️

mnkiefer commented on lib/linter/linter.js:
```diff
@@ -818,6 +822,38 @@ function stripUnicodeBOM(text) {
+        slots.times.passes[passIndex][type][key] = slots.times.passes[passIndex][type][key] || { total: 0 };
+        slots.times.passes[passIndex][type][key].total += time;
+    } else if (type !== "rules") {
+        slots.times.passes[passIndex][type].total = time;
```
Sorry for the typo, indeed this should be added, same as the first block! 👍

mdjermanovic commented:
Should we maybe also document the Stats object?

It is intended for use by custom formatters and integrations, but there's no explanation of its structure. Perhaps we could add a new page that describes the Stats object, and add links to it from Custom Formatters and Node.js API Reference pages.

aladdin-add commented:
Tried this option on my computer and it didn't output anything:
```sh
➜  eslint git:(pr17850) node ./bin/eslint.js --stats ./eslint.config.js
(node:58409) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)
```

am i missing something?

mdjermanovic commented on lib/shared/types.js:
```diff
@@ -189,11 +189,44 @@ module.exports = {};
  * @property {number} warningCount Number of warnings for the result.
  * @property {number} fixableErrorCount Number of fixable errors for the result.
  * @property {number} fixableWarningCount Number of fixable warnings for the result.
+ * @property {Stats} Stats The performance statistics collected with the `stats` flag.
```
```suggestion
 * @property {Stats} [stats] The performance statistics collected with the `stats` flag.
```

Lowercase `stats` is the property name. `[]` because it's optional (it may or may not be present depending on whether the option was used).

mdjermanovic commented on lib/shared/types.js:
```diff
@@ -189,11 +189,44 @@ module.exports = {};
+ * @property {ParseTime} parse The parse object containing all parse time information.
+ * @property {Record<string, RuleTime>} rules The rules object containing all lint time information for each rule.
+ * @property {FixTime} fix The parse object containing all fix time information.
+ * @property {string} total The total time that is spent on (parsing, fixing, linting) a file.
```
```suggestion
 * @property {number} total The total time that is spent on (parsing, fixing, linting) a file.
```

mdjermanovic commented on lib/shared/types.js:
```diff
@@ -189,11 +189,44 @@ module.exports = {};
+ */
+/**
+ * @typedef {Object} FixTime
+ * @property {number} total The total time that is spent on a fix pass.
```
```suggestion
 * @property {number} total The total time that is spent on applying fixes to the code.
```

"fix pass" may sound like the total time spent in a pass, while this is specifically time spent on applying fixes.

mdjermanovic commented on lib/shared/types.js:
```diff
@@ -189,11 +189,44 @@ module.exports = {};
+ */
+
+/**
+ * Performance Times for each ESLint pass
+ * @typedef {TimePass[]} Time passes
```
```suggestion
 * Performance Times for each ESLint pass
 * @typedef {Object} Times
 * @property {TimePass[]} passes Time passes
```

mdjermanovic commented:
@aladdin-add it doesn't output anything by itself, just adds more data to lint results. Try with `-f json`.

We should clarify this in the CLI docs.

mnkiefer commented:
> Should we maybe also document the Stats object?
> 
> It is intended for use by custom formatters and integrations, but there's no explanation of its structure. Perhaps we could add a new page that describes the Stats object, and add links to it from Custom Formatters and Node.js API Reference pages.

I agree, this needs to be mentioned somewhere central so it is visible and can be easily referenced from different places in the documentation. Now that I think about it, it is also at least somewhat related to the `TIMING` section (i.e. custom rules).

Where (dir/file) would be the best place to put such Stats "how-to" information?

mnkiefer commented:
@nzakas: I was wondering, once flat ESLint (default) is released and the *stats* feature is out, could I contribute a post to [Case studies](https://eslint.org/blog/category/case-studies/) on this feature on how we came up with it, its benefits and our decision to generalize it and make it available to all plugin developers via ESLint?

mdjermanovic commented:
> Should all of this be under a header such as _Performance_ in the docs? Otherwise, where (dir/file) would be a good place to put such Stats "how-to" information? This section would also explain having to add output options to the CLI [as already mentioned](https://github.com/eslint/eslint/pull/17850#issuecomment-1980606408).

Since the primary use case is custom formatters, which are documented under `extend`, the new document could be `docs/src/extend/stats.md`, with the title `Stats Data`.

We would have links to it from Custom Formatters, Node.js API Reference, Custom Rules (in the Profile Rule Performance section), and Command Line Interface Reference (in the new `--stats` section that we should add).

mnkiefer commented:
> Since the primary use case is custom formatters, which are documented under `extend`, the new document could be `docs/src/extend/stats.md`, with the title `Stats Data`.

Added the [`Stats Data`](https://github.com/eslint/eslint/pull/17850/commits/536bc5b9981843d77cb45e6447f81aee8aecfc69#diff-c8a35e7ac5866cf15afc9a0401df9f1edf1be41ce00e608163661e2c14f9740fR1-R7) page 👍
 
> We would have links to it from Custom Formatters, Node.js API Reference, Custom Rules (in the Profile Rule Performance section), and Command Line Interface Reference (in the new `--stats` section that we should add).

Added links from pages:
- [CLI](https://github.com/eslint/eslint/pull/17850/commits/536bc5b9981843d77cb45e6447f81aee8aecfc69#diff-80937556f9a8e68352718255c20699fce061fa72760504c6accc3fc6b3b0613aR815-R818)
- [Custom formatters](https://github.com/eslint/eslint/pull/17850/commits/536bc5b9981843d77cb45e6447f81aee8aecfc69#diff-a1e0ac605cf6c4425727760f6811749e3d1868ee3b60d7da0f529c1d668af0c1R101)
- [Custom Rules](https://github.com/eslint/eslint/pull/17850/commits/536bc5b9981843d77cb45e6447f81aee8aecfc69#diff-dc3f44ef800795b890a57c8dd95c6ce26ba8a47a4551d7a05d7a5cd9cf016597R942)
- [Node.js API](https://github.com/eslint/eslint/pull/17850/commits/536bc5b9981843d77cb45e6447f81aee8aecfc69#diff-6babf40ac9ef9dedc10f6dd89574fca30aa1adc48d09b7f446effd151f67fb9a)

Please feel free to edit/expand wherever the text is unclear or you feel some details are missing.

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2166,15 +2259,46 @@ class Linter {
          */
+        const slots = internalSlotsMap.get(this);
+
+        // Remove lint times from the last run.
+        if (stats) {
+            delete slots.times;
+        }
```
I think we should reset `slots.fixPasses` as well:

```js
"use strict";

const { ESLint } = require("eslint");
const eslint = new ESLint({
    stats: true,
    fix: true,
    overrideConfigFile: true,
    overrideConfig: {
        rules: {
            "no-extra-parens": 2
        }
    }
});

(async () => {
    await eslint.lintText("((foo))"); // 2 fix passes
    await eslint.lintText("foo;"); // "TypeError: Cannot read properties of undefined (reading 'parse')" - reads `slots.times.passes[2]["parse"]`
})();

```

mnkiefer replied:
Done

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,116 @@
+
+## ◆ Stats type
+
+The `Stats` value is the timing information of each lint run. The `stats` property of the [LintResult] type contains it. It has the following properties:
```
```suggestion
The `Stats` value is the timing information of each lint run. The `stats` property of the [LintResult](../integrate/nodejs-api#-lintresult-type) type contains it. It has the following properties:
```

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,116 @@
+  The number of times ESLint has applied at least one fix after linting.
+* `times` (`TimePass[]`)<br>
+  The times spent on (parsing, fixing, linting) a file, where the linting refers to the timing information for each rule.
+    * `TimePass` (`{ parse: ParseTime, rules: Record<string, RuleTime> fix: FixTime }`)<br>
```
```suggestion
    * `TimePass` (`{ parse: ParseTime, rules: Record<string, RuleTime>, fix: FixTime, total: number }`)<br>
```

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,116 @@
+  The times spent on (parsing, fixing, linting) a file, where the linting refers to the timing information for each rule.
+    * `TimePass` (`{ parse: ParseTime, rules: Record<string, RuleTime> fix: FixTime }`)<br>
+    An array containing the times spent on (parsing, fixing, linting)
+    * `ParseTime` (`{ total: number }`)<br>
+      The total time that is spent when parsing a file.
+    * `RuleTime` (`{ total: number }`)<br>
+      The total time that is spent on a rule.
+    * `FixTime` (`{ total: number }`)<br>
+      The total time that is spent on applying fixes to the code.
```
```suggestion
      * `ParseTime` (`{ total: number }`)<br>
      The total time that is spent when parsing a file.
      * `RuleTime` (`{ total: number }`)<br>
      The total time that is spent on a rule.
      * `FixTime` (`{ total: number }`)<br>
      The total time that is spent on applying fixes to the code.
```

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,116 @@
+* `times` (`TimePass[]`)<br>
+  The times spent on (parsing, fixing, linting) a file, where the linting refers to the timing information for each rule.
+    * `TimePass` (`{ parse: ParseTime, rules: Record<string, RuleTime> fix: FixTime }`)<br>
+    An array containing the times spent on (parsing, fixing, linting)
```
```suggestion
    An object containing the times spent on (parsing, fixing, linting)
```

mdjermanovic commented on docs/src/extend/stats.md:
````diff
@@ -0,0 +1,116 @@
+    console.error(error);
+});
+```
+
+---
````
```suggestion
```

mdjermanovic commented on docs/src/extend/stats.md:
````diff
@@ -0,0 +1,116 @@
+Run ESLint with the *stats* option (`--stats`) and outputting to JSON via a [built-in formatter](../use/formatters/) (`-f json`):
+
+```sh
+eslint file-to-fix.js --stats -f json
````
Since the file is named "file-to-fix", it might be more representative to run eslint with `--fix` and get an array with multiple elements in `passes`?

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,116 @@
+---
+title: Stats Data
+eleventyNavigation:
+    key: Stats Data
+    parent: extend eslint
+    title: Stats Data
```
```suggestion
    key: stats data
    parent: extend eslint
    title: Stats Data
    order: 6
```

mnkiefer commented on lib/linter/timing.js:
```diff
@@ -128,21 +130,26 @@ module.exports = (function() {
-            return result;
+            const tdiff = endTime(t);
+
+            data[key] += tdiff;
```
Done - wrapped it [here](https://github.com/eslint/eslint/pull/17850/files#diff-126a649c1db33de2cfe67b418435b10d45fc310143547e334f7be9a1a73c0901R145-R151).

mdjermanovic commented on docs/src/extend/stats.md:
````diff
@@ -0,0 +1,132 @@
+Note, that for the simple example above, the sum of all rule times should be directly comparable to the first column of the TIMING output. Running the same command with `TIMING=all`, you can verify this:
+
+```bash
+$ TIMING:all eslint file-to-fix.js --fix --stats -f json
````
```suggestion
$ TIMING=all npx eslint file-to-fix.js --fix --stats -f json
```

mdjermanovic commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,132 @@
+
+(async function main() {
+    // 1. Create an instance.
+    const eslint = new ESLint({ stats: true });
```
```suggestion
    const eslint = new ESLint({ stats: true, fix: true });
```

mnkiefer commented on tests/lib/eslint/eslint.js:
```diff
@@ -5060,6 +5060,73 @@ describe("ESLint", () => {
+            assert.strictEqual(isNumber(results[0].stats.times.passes[0].fix.total), true);
+            assert.strictEqual(isNumber(results[0].stats.times.passes[1].fix.total), true);
+            assert.strictEqual(isNumber(results[0].stats.times.passes[0].total), true);
+            assert.strictEqual(isNumber(results[0].stats.times.passes[1].total), true);
```
Done [here](https://github.com/eslint/eslint/pull/17850/commits/e954f444f066b9bb97fbed3413a58afe83c9b59b#diff-16e358f44e63489038dac1466edcc3860de362b72d78279cec28d9ad2209b5e0R5114)

mnkiefer commented on tests/lib/eslint/eslint.js:
```diff
@@ -5060,6 +5060,73 @@ describe("ESLint", () => {
+         * @returns {boolean} `true` if the number is a float, `false` otherwise.
+         */
+        function isNumber(n) {
+            return Number(n) === n && n % 1 !== 0;
```
Changed this to returning `typeof n === "number"`[here](https://github.com/eslint/eslint/pull/17850/files#diff-16e358f44e63489038dac1466edcc3860de362b72d78279cec28d9ad2209b5e0R5071)

mdjermanovic commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -133,6 +133,8 @@ The `ESLint` constructor takes an `options` object. If you omit the `options` ob
   Default is `null`. Ignore file patterns to use in addition to config ignores. These patterns are relative to `cwd`.
 * `options.passOnNoPatterns` (`boolean`)<br>
   Default is `false`. When set to `true`, missing patterns cause the linting operation to short circuit and not report any failures.
+* `options.stats` (`boolean`)<br>
+  Default is `false`. When set to `true`, additional statistics are added to the lint results (see [Stats type](../extend/stats#-stats-type)).
```
Can you move this into the Linting section below, as this option isn't related to File Enumeration?

mdjermanovic commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -365,6 +367,8 @@ The `LintResult` value is the information of the linting result of each file. Th
   The number of warnings. This includes fixable warnings.
 * `output` (`string | undefined`)<br>
   The modified source code text. This property is undefined if any fixable messages didn't exist.
+* `stats` (`Stats | undefined`)<br>
+  The [Stats](../extend/stats#-stats-type) object. This contains the lint performance statistics collected with the `stats` option.
 * `source` (`string | undefined`)<br>
   The original source code text. This property is undefined if any messages didn't exist or the `output` property exists.
```
```suggestion
* `output` (`string | undefined`)<br>
  The modified source code text. This property is undefined if any fixable messages didn't exist.
* `source` (`string | undefined`)<br>
  The original source code text. This property is undefined if any messages didn't exist or the `output` property exists.
* `stats` (`Stats | undefined`)<br>
  The [Stats](../extend/stats#-stats-type) object. This contains the lint performance statistics collected with the `stats` option.
```

Just to keep `output` and `source` together as they're logically related.

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2189,6 +2314,13 @@ class Linter {
+                const passIndex = slots.times.passes.length - 1;
+
+                slots.times.passes[passIndex].total = tTotal;
+            }
```
I think we could fix this directly in that branch:

```diff
if (fixedResult.fixed) {
+   let tTotal;
+
+   if (stats) {
+       tTotal = startTime();
+   }

    fixedResult.messages = this.verify(currentText, config, options);

+   if (stats) {
+       storeTime(0, { type: "fix" }, slots);
+       slots.times.passes.at(-1).total = endTime(tTotal);
+   }
}
```

mnkiefer replied:
Thanks for the fix, @mdjermanovic! 🚀  I can confirm that it works now and that the 11th pass contains the previously missing `fix` and `total` properties. 👍🏼

mdjermanovic reviewed (dismissed):
LGTM, thanks! Leaving open for @nzakas to verify again as this has had a lot of changes since the last approval.

nzakas commented on docs/src/extend/stats.md:
````diff
@@ -0,0 +1,132 @@
+}
+```
+
+Run ESLint with the *stats* option (`--stats`) and outputting to JSON via a [built-in formatter](../use/formatters/) (`-f json`):
````
```suggestion
Run ESLint with `--stats` and output to JSON via the built-in [`json` formatter](../use/formatters/):
```

nzakas commented on docs/src/extend/stats.md:
```diff
@@ -0,0 +1,132 @@
+
+### API Usage
+
+Similarly to the example for the [CLI usage](#cli-usage), we can achieve the same thing using the Node.js API, turning on the *stats* option (`stats: true`) as follows:
```
```suggestion
You can achieve the same thing using the Node.js API by passing`stats: true` as an option to the `ESLint` constructor. For example:
```

nzakas commented on lib/linter/linter.js:
```diff
@@ -2081,6 +2160,14 @@ class Linter {
+        return internalSlotsMap.get(this).times ?? { passes: [] };
+    }
+
+    getFixPasses() {
```
Could we call this `getFixPassCount()` to indicate that this is a number rather than an array?

nzakas commented on docs/src/integrate/nodejs-api.md:
Can you also add docs for `Linter#getTimings()` and `Linter#getFixPasses()`?

nzakas reviewed (changes requested):
Just left a few notes of things to clean up before merging.

mnkiefer commented on lib/linter/linter.js:
```diff
@@ -2081,6 +2160,14 @@ class Linter {
+    getFixPasses() {
+        return internalSlotsMap.get(this).fixPasses ?? 0;
+    }
+
```
Done 👍

nzakas reviewed (approved):
LGTM. Once again, thank you for your hard work on such a detailed feature. We really appreciate it. 🙏
